README.md
#### Preface

This is a simple, unofficial library for creating Stackdriver traces. Google
has an [alpha library](https://github.com/GoogleCloudPlatform/cloud-trace-nodejs)
that instruments a variety of common modules, which you may prefer to use
instead.

## Authentication
If running on Google Cloud Platform, authentication can happen automatically
if you have the `https://www.googleapis.com/auth/trace.append` or `https://www.googleapis.com/auth/cloud-platform`
(a.k.a. `cloud-platform`) OAuth scope enabled on the resource where you're
running this library. Otherwise you will need to provide credentials:

```js
var projectId = process.env.GCLOUD_PROJECT; // e.g. 'grape-spaceship-123'

var tracing = require('gcloud-trace')({
  projectId: projectId,

  // The path to your key file:
  keyFilename: '/path/to/keyfile.json'

  // Or the contents of the key file:
  credentials: require('./path/to/keyfile.json')
});
```

## Example

```js
var config = {
	projectId: 'my-project'
};
var tracing = require('gcloud-trace')(config);

function myfn(req, res, next) {
	var span0 = tracing.startRootSpan('name'); // creates a trace and root-level span
	// Do some work

	// Optionally create nested spans with or without labels
	var span1 = span0.startSpan('child span 1', {key1: 'value 1'});
	// Do some work

	var span2 = span1.startSpan('child span 1-1');
	// Do some other work

	span2.end(); // optional; if omitted will end when span1 or trace/span0 ends
	span1.end(); // optional; if omitted will end when trace/span0 ends
	span0.end(); // required
}
```

## API

### Class: `Tracing`

#### require('gcloud-trace')(config)
* `config.writeInterval <Number>` Maximum write interval in seconds.
* `config.maxBufferSize <Number>` Maximum buffer size before a write is forced.
* Return: `tracing` instance.

#### tracing.startRootSpan(name[, labels])
* `name <String>` Name for the span.
* `labels <Object>` Optional labels to apply to the span.
* Return: `<Span>`

This is what you will typically use to create traces.

Creates a trace with a root span with the provided name and optional labels.
The `end` method on the returned span is augmented to also end the trace.
(Calling `end` on a span created with `trace.startSpan()` does not end the
trace.) The returned span also has a `cancel` method (see
[`Trace.cancel`](#tracecancel)).

Note that this root-level span's name is used as the URI in the Stackdriver
Trace web interface trace list.

#### tracing.startTrace()
* Return: `<Trace>`

Creates an empty trace.

#### tracing.getTrace(traceId, callback)
* `traceId <String>` Trace ID to retrieve.
* `callback <Function>` The callback function.
  * `callback.err <?Error>` An error returned while making this request.
  * `callback.trace <Trace>` The trace.

The Stackdriver Trace API allows patching of traces at any time. You can thus
use this to retrieve an existing trace and modify existing or add new spans to
it. For example, you could send a traceId to an RPC server and continue tracing
on the other server.

### Class: `Trace`

#### trace.startSpan(name[, labels])
* `name <String>` Name for the span.
* `labels <Object>` Optional labels to apply to the span.
* Return: `<Span>`

Starts a new root-level span on this trace.

#### trace.end()

Ends all of the trace's spans and queues it for writing.

#### trace.cancel()
*Added in 1.1.0*

Cancels the trace. All methods remain intact on the trace and its spans, but
calling `end` will not send the trace to the server.

### Class: `Span`

#### span.end([labels])
* `labels <Object>` Optional key-value label pairs to apply to the span.

Ends this span.

#### span.startSpan(name[, labels])
* `name <String>` Name for the span.
* `labels <Object>` Optional labels to apply to the span.
* Return: `<Span>`

Starts a new child span on this span.

#### Enum Span.SPAN_KIND
* `UNSPECIFIED`
* `RPC_SERVER`
* `RPC_CLIENT`

These don't seem to have an effect on display in the GCP console, but you can
set the span kind to these values if you wish:
`span.kind = Span.SPAN_KIND.RPC_SERVER`. See [docs](https://cloud.google.com/trace/api/reference/rest/v1/projects.traces#SpanKind).

## Hooks

### Express
*Added in 1.1.0*

```js
var tracing = require('gcloud-trace')({projectId: 'my-project'});
var app = express();
app.use(tracing.Express());
app.use('/path', function (req, res) {
  // Use req.trace.startSpan to trace operations within a request:
  var span = req.trace.startSpan('doing work');
  // do work...
  span.end(); // optional; will get closed when the request ends.
});
```

The `tracing.Express` middleware traces requests end-to-end with a root span,
which is automatically labeled with the method, URL, hostname, status code and
response size. That span is exposed as `req.trace` if you want to add child
spans in your request handlers or other middleware.

Use the `config` to set rate limiting on tracing (default is unlimited):

```js
app.use(tracing.Express({
  maxRPS: 100 // trace up to 100 requests per second
}));
```

If you want to skip tracing for a particular route or under particular cases,
call `req.trace.cancel()`.

The `o=TRACE_TRUE` suffix of the `X-Cloud-Trace-Context` header can be used to
force a request to be traced or not traced (see [docs](https://cloud.google.com/trace/docs/faq#how_do_i_force_a_request_to_be_traced)).
If provided, the `TRACE_ID` and `SPAN_ID` fields are used as well. Valid headers
patterns include:

* `X-Cloud-Trace-Context: o=1` trace and use auto-generated IDs
* `X-Cloud-Trace-Context: b75dc0042a82efcb6b0a194911272926` use specified
  traceId and auto-generated spanId. Use default tracing decision (e.g. `maxRPS`
  setting).
* `X-Cloud-Trace-Context: b75dc0042a82efcb6b0a194911272926/1258215` use
  specified traceId, and start counting from specified spanId. Use default
  tracing decision.
* `X-Cloud-Trace-Context: b75dc0042a82efcb6b0a194911272926/1258215;o=1` use
  specified traceId and spanId, and force tracing.

This is useful for continuing traces across servers/services.

Note that Google specifies `SPAN_ID` as a 64-bit integer, which is not
necessarily safe in JavaScript. gcloud-trace will only use the provided
`SPAN_ID` if it is a safe JavaScript integer; otherwise it will default to 1.
(GCP load balancers frequently generate unsafe JavaScript integers.)

## Tips

### Labels
Google uses several special labels present on root spans to display traces
nicely. Some of them are documented [here](https://cloud.google.com/trace/docs/viewing-details#view_the_timeline);
the rest listed below are based on observations:

* `"trace.cloud.google.com/http/status_code": "200"`
  * Displayed on the trace details page; used for filtering on trace list page.
* `"trace.cloud.google.com/http/method": "POST"`
  * Displayed on the trace list and details pages.
* `"trace.cloud.google.com/http/url": "http://test.com/path/"` (or relative URL)
  * Displayed at the top of the trace details pages.
* `"trace.cloud.google.com/http/host": "hostname"`
  * This is listed in the above-linked docs, but doesn't seem to be handled
  specially. (Only shows up in the sidebar.)
* `"trace.cloud.google.com/http/response/size": "123456"`
  * This is listed in the above-linked docs, but doesn't seem to be handled
  specially. (Only shows up in the sidebar.)
* `"trace.cloud.google.com/gae/app/module": "service name"`
  * Displayed as the "Service" in Trace Details.
* `"trace.cloud.google.com/gae/request_log_id": "12ab8cd8fh9ac"`
  * Used to connect the "show logs" button in the trace viewer to the
  Stackdriver logging interface/viewer. This seems to be limited to querying
  appengine logs based on the log's `request_id`.

Also note that the root-level span's name is used as the URI in the trace list.

### Effects on node.js process

Write are batched and dispatched at the `writeInterval` interval. The batching
timer is weakly referenced; it will not keep the node process alive after a
shutdown has started, but consequently writes may be lost on shutdown.


CHANGELOG.md
1.3.3
============
### Source
* Update dependency.

1.3.2
============
### Feature
* Include hostname in Express hook.

1.3.1
============
### Bugfix
* Only use SPAN\_ID if it is a safe JS integer.

1.3.0
============
### Feature
* Use TRACE\_ID and SPAN\_ID in Express hook if provided in request header.

1.2.1
============
### Bug fixes
* Ignore `X-Cloud-Trace-Context` header if it lacks the `o=\d` suffix. (GCP
load balancers can add this header without the suffix.)

1.2.0
============
### Features
* Add `span.cancel` method.

1.1.0
============
### Features
* Add `trace.cancel` method.
* Add `tracing.Express` middleware for tracing Express requests.

1.0.0
============
Initial release.


test/e2e.js
var Tracing = require('..');
var tracing = Tracing({projectId: process.env.GCLOUD_PROJECT});
var assert = require('assert');

// Keep node alive while the tests run.
setTimeout(function () {}, tracing.writeInterval * 1000 * 4);

(function simple() {
	var span0 = tracing.startRootSpan('root span');
	assert.ok(span0.startTime);
	assert.equal(span0.name, 'root span');
	assert.deepEqual(span0.labels, {});

	setTimeout(function () {
		var span1 = span0.startSpan('child span 1', {key: 'val'});
		assert.equal(span1.parentSpanId, span0.spanId);
		assert.ok(span1.startTime);
		assert.equal(span1.name, 'child span 1');
		assert.deepEqual(span1.labels, {key: 'val'});

		setTimeout(function () {
			span1.end();
			assert.ok(span1.endTime);
			var span2 = span0.startSpan('child span 2');

			setTimeout(function () {
				span0.end();
				assert.ok(span2.endTime);
				assert.ok(span0.endTime);

			}, 50);
		}, 50);
	}, 50);
})();

(function simpleWithKind() {
	var span0 = tracing.startRootSpan('root span with kinds');

	setTimeout(function () {
		var span1 = span0.startSpan('RPC client span');
		span1.kind = Tracing.Span.SPAN_KIND.RPC_CLIENT;

		var span11 = span0.startSpan('RPC client span 2');
		span11.kind = Tracing.Span.SPAN_KIND.RPC_CLIENT;

		setTimeout(function () {
			var span2 = span1.startSpan('RPC server span');
			span2.kind = Tracing.Span.SPAN_KIND.RPC_SERVER;

			setTimeout(function () {
				span0.end();
			}, 50);
		}, 50);
	}, 50);
})();

(function specialLabels() {
	var labels = {
		'trace.cloud.google.com/http/status_code': '200',
		'trace.cloud.google.com/http/method': 'POST',
		'trace.cloud.google.com/http/url': 'http://test.com/path/',
		'trace.cloud.google.com/http/host': 'hostname',
		'trace.cloud.google.com/http/response/size': '123456',
		'trace.cloud.google.com/gae/app/module': 'service name',
		'trace.cloud.google.com/gae/request_log_id': 'sl5ei9fdplskw'
	};
	var span0 = tracing.startRootSpan('root span with labels', labels);

	setTimeout(function () {
		span0.end();
	}, 50);
})();

(function patchExisting() {
	var span0 = tracing.startRootSpan('original existing span');

	setTimeout(function () {
		span0.end();

		setTimeout(function () {
			tracing.getTrace(span0._trace.traceId, function (err, trace) {
				assert.ifError(err);
				assert(trace instanceof Tracing.Trace);
				assert(trace.spans[0] instanceof Tracing.Span);
				trace.spans[0].name = 'patched existing span';
				trace.end();
			});
		}, tracing.writeInterval * 1000 * 2.2);

	}, 50);
})();


(function cancelSpan() {
	var span0 = tracing.startRootSpan('root span with cancelled span');
	var span1 = span0.startSpan('cancelled span');
	setTimeout(function () {
		span1.cancel();
		span0.end();
	}, 50);
})();


test/hooks/express/express.js
var Tracing = require('../../..');
var tracing = Tracing({projectId: process.env.GCLOUD_PROJECT});
var assert = require('assert');
var express = require('express');
var request = require('request');

// Keep node alive while the tests run.
setTimeout(function () {}, tracing.writeInterval * 1000 * 3);

(function simple() {
	var app = express();
	app.use(tracing.Express({}));
	app.get('/', function (req, res) {
		assert.ok(req.trace);
		setTimeout(function () {
			res.send({message: 'hey!'});
		}, 150);
	});
	app.get('/foo/bar', function (req, res) {
		setTimeout(function () {
			var childSpan = req.trace.startSpan('child span');
			setTimeout(function () {
				childSpan.end();
				res.send({message: 'a deeper url'});
			}, 250);
		}, 50);
	});
	var server = app.listen(8080, function () {
		request('http://localhost:8080/', function (err, res, body) {
			assert.ifError(err);
			request('http://localhost:8080/foo/bar', function (err, res, body) {
				assert.ifError(err);
				server.close();
			});
		});
	});
})();

(function rateLimited() {
	var app = express();
	app.use(tracing.Express({
		maxRPS: 1
	}));
	app.get('/', function (req, res) {
		assert.ok(req.trace);
		setTimeout(function () {
			res.send({message: 'hey!'});
		}, 150);
	});
	var server = app.listen(8081, function () {
		for (var i = 0; i < 10; i++) {
			request('http://localhost:8081/foo/bar', function (err, res, body) {
				assert.ifError(err);
			});
		}
		setTimeout(function () { server.close(); }, 500);
	});
})();

(function ctHeader() {
	var app = express();
	app.use(tracing.Express({
		maxRPS: 0
	}));
	app.get('/', function (req, res) {
		assert.ok(req.trace);
		setTimeout(function () {
			res.send({message: 'hey!'});
		}, 150);
	});
	var server = app.listen(8082, function () {
		request({
			url: 'http://localhost:8082/foo/bar?shouldbelogged',
			headers: {
				'X-Cloud-Trace-Context': '105445aa7843bc8bf206b120001000/0;o=1'
			}
		}, function (err, res, body) {
			assert.ifError(err);
			server.close();
		});
	});
})();


src/span.js
function Span(trace, name, labels, parentSpan) {
	this.spanId = trace.getNextSpanId();
	this.kind = Span.SPAN_KIND.UNSPECIFIED;
	this.name = name;
	this.startTime = getTimestamp();
	this.labels = labels || {};
	this.parentSpanId = undefined;
	if (parentSpan) {
		parentSpan._spans.push(this);
		this.parentSpanId = parentSpan.spanId;
	}

	this._parentSpan = parentSpan;
	this._spans = [];
	this._trace = trace;
	trace.spans.push(this);
}

Span.SPAN_KIND = {
	UNSPECIFIED: 'SPAN_KIND_UNSPECIFIED',
	RPC_SERVER: 'RPC_SERVER',
	RPC_CLIENT: 'RPC_CLIENT'
};

/**
 * Ends a span and all of its child spans. Idempotent.
 *
 * @param {?object} labels - Labels to add to span. If provided, labels will
 * only be applied to this span, not its children.
 * @param {?boolean} shallow - Do not end child spans.
 */
Span.prototype.end = function end(labels, shallow) {
	if (!shallow) {
		this._spans.forEach(function (span) {
			span.end();
		});
	}
	if (!this.endTime) this.endTime = getTimestamp();
	if (labels) {
		for (var k in labels) {
			this.labels[k] = labels[k];
		}
	}
};

/**
 * Removes a span from the trace. And child spans on this trace will be
 * escalated so that their new parent is this span's parent.
 */
Span.prototype.cancel = function cancel() {
	// Remove from trace.
	var idx = this._trace.spans.indexOf(this);
	if (idx !== -1) this._trace.spans.splice(idx, 1);
	// Remove from parent span if present.
	if (this._parentSpan) {
		idx = this._parentSpan._spans.indexOf(this);
		if (idx !== -1) this._parentSpan._spans.splice(idx, 1);
	}
	// Escalate child spans in the graph.
	var thisParentSpanId = this.parentSpanId;
	this._spans.forEach(function (span) {
		if (span.parentSpanId) span.parentSpanId = thisParentSpanId;
	});
};

Span.prototype.toObject = function toObject() {
	return {
		spanId: this.spanId,
		kind: this.kind,
		name: this.name,
		startTime: this.startTime,
		endTime: this.endTime,
		labels: this.labels,
		parentSpanId: this.parentSpanId
	};
};

/**
 * Starts a child span on this span.
 *
 * @param {string} name - Name of span.
 * @param {?object} labels - Labels to add to span.
 */
Span.prototype.startSpan = function startSpan(name, labels) {
	var child = new Span(
		this._trace,
		name,
		labels,
		this
	);
	return child;
};

function getTimestamp() {
	return (new Date()).toISOString();
}

module.exports = Span;


src/index.js
var common = require('@google-cloud/common');
var util = require('util');
var extend = require('extend');
var Trace = require('./trace');
var Span = require('./span');
var createExpressTracer = require('./hooks/express');

var defaultConfig = {
	writeInterval: 5,
	maxBufferSize: 1000
};

function Tracing(options) {
	if (!(this instanceof Tracing)) {
		extend(options, defaultConfig);
		options = common.util.normalizeArguments(this, options);
		return new Tracing(options);
	}

	var config = {
		baseUrl: 'https://cloudtrace.googleapis.com/v1',
		scopes: [
			'https://www.googleapis.com/auth/trace.append',
			'https://www.googleapis.com/auth/trace.readonly'
		],
		packageJson: require('../package.json')
	};

	common.Service.call(this, config, options);

	/**
	 * Closed traces ready to be written.
	 */
	this._buffer = [];

	this._writeQueued = false;

	this.writeInterval = options.writeInterval;
	this.maxBufferSize = options.maxBufferSize;
}

util.inherits(Tracing, common.Service);

/**
 * Gets an existing trace.
 *
 * @param {string} traceId - ID of the trace to retrieve
 * @param {function} callback - The callback function.
 * @param {?error} callback.err - An error returned while making this request.
 * @param {Trace} callback.trace - The full trace.
 */
Tracing.prototype.getTrace = function get(traceId, callback) {
	var _this = this;
	var uri = '/traces/' + traceId;
	this.request({
		uri: uri
	}, function (err, trace) {
		if (err) return callback(err);
		var result = new Trace(_this, trace.traceId);
		var spans = {};
		for (var i = 0; i < trace.spans.length; i++) {
			var spanI = trace.spans[i];
			var spanIR = new Span(
				result,
				spanI.name,
				spanI.labels);
			spanIR.startTime = spanI.startTime;
			spanIR.endTime = spanI.endTime;
			spanIR.kind = spanI.spanKind;
			spans[spanI.spanId] = spanIR;
		}
		result.spans = trace.spans.map(function (span) {
			var spanIR = spans[span.spanId];
			if (span.parentSpanId) {
				spanIR.parentSpanId = span.parentSpanId;
				spans[spanIR.parentSpanId]._spans.push(spanIR);
			}
			return spanIR;
		});
		callback(null, result);
	});
};

/**
 * Starts a trace and a root span. Calling 'end' on the returned root span will
 * also end the trace. Calling 'cancel' will cancel the trace.
 */
Tracing.prototype.startRootSpan = function startRootSpan(name, labels) {
	var trace = this.startTrace();
	var rootSpan = trace.startSpan(name, labels);
	rootSpan.end = trace.end.bind(trace);
	rootSpan.cancel = trace.cancel.bind(trace);
	return rootSpan;
};

/**
 * Starts a trace.
 */
Tracing.prototype.startTrace = function startTrace() {
	return new Trace(this);
};

/**
 * Internal.
 * Queues a completed trace for writing.
 */
Tracing.prototype._queueForWrite = function _queueForWrite(trace) {
	this._buffer.push(trace);
	if (this._buffer.length > this.maxBufferSize) {
		setImmediate(this._write.bind(this));
	} else if (!this._writeQueued) {
		this._writeQueued = true;
		setTimeout(this._write.bind(this), this.writeInterval * 1000).unref();
	}
};

/**
 * Internal.
 * Writes buffer to API.
 */
Tracing.prototype._write = function _write() {
	this._writeQueued = false;
	var traces = this._buffer;
	this._buffer = [];
	var body = JSON.stringify({traces: traces});

	this.request({
		method: 'PATCH',
		uri: '/traces',
		body: body
	}, function (err) {
		if (err) console.log(err);
	});
};

Tracing.prototype.Express = function (config) {
	return createExpressTracer(this, config);
};

Tracing.Trace = Trace;
Tracing.Span = Span;

module.exports = Tracing;


src/trace.js
var Span = require('./span');
var crypto = require('crypto');

function Trace(agent, traceId) {
	this._agent = agent;
	this._spanIdInc = 1;
	this.projectId = agent.projectId;
	this.traceId = traceId || hex(32);
	this.spans = [];
}

Trace.prototype.getNextSpanId = function () {
	return (this._spanIdInc++).toString();
};

/**
 * Starts a root-level span on this trace.
 */
Trace.prototype.startSpan = function startSpan(name, labels) {
	var span = new Span(
		this,
		name,
		labels
	);
	return span;
};

/**
 * Ends this trace and all of its spans, then queues it for writing.
 */
Trace.prototype.end = function end() {
	this.spans.forEach(function (span) {
		// Spans created with startRootSpan have an augmented 'end' method.
		// Need to explicitly call the prototype method to avoid infinite
		// recursion.
		Span.prototype.end.call(span, null, true);
	});
	this._agent._queueForWrite(this.toObject());
};

/**
 * Cancels a trace so it will not be send to the server.
 */
Trace.prototype.cancel = function cancel() {
	this.end = noop;
};

/**
 * Formats this Trace for writing (i.e. removes private members).
 */
Trace.prototype.toObject = function toObject() {
	return {
		projectId: this.projectId,
		traceId: this.traceId,
		spans: this.spans.map(function (s) { return s.toObject(); })
	};
};

/**
 * Generates a random hex string of the provided length.
 */
function hex(n) {
	return crypto.randomBytes(Math.ceil(n / 2)).toString('hex').slice(0, n);
}

function noop() {}

module.exports = Trace;


src/hooks/express.js
var XCTC_HEADER = 'x-cloud-trace-context';
var TRACE_ID_REGEX = /^([0-9a-f]{32})/;
var SPAN_ID_REGEX = /\/(\d+)/;
var O_SUFFIX_REGEX = /o=\d$/;
var O1_SUFFIX_REGEX = /o=1$/;

module.exports = function createExpressTracer(tracing, config) {
	var shouldTrace = (function () {
		if (config && typeof config.maxRPS === 'number') {
			var maxRPS = config.maxRPS;
			var requests = 0;
			setInterval(function () {
				requests = 0;
			}, 1000).unref();
			return function shouldTrace(req) {
				var xctc = req.get(XCTC_HEADER);
				if (xctc && O_SUFFIX_REGEX.test(xctc)) {
					return O1_SUFFIX_REGEX.test(xctc);
				}
				return requests++ < maxRPS;
			};
		} else {
			return function shouldTrace(req) {
				var xctc = req.get(XCTC_HEADER);
				if (xctc) return O1_SUFFIX_REGEX.test(xctc);
				return true;
			};
		}
	})();

	var hostname = require("os").hostname();

	/**
	 * Express middleware. Adds a root span to the request, labeled with basic
	 * information about the request. Trace subspans using:
	 *
	 * ```
	 * var span = req.trace.startSpan('name'[, labels]);
	 * ...
	 * span.end();
	 * ```
	 */
	return function traceExpress(req, res, next) {
		var method = req.method;
		var url = (req.baseUrl || '') + (req.url || '-');
		var labels = {
			'trace.cloud.google.com/http/method': method,
			'trace.cloud.google.com/http/url': url,
			'trace.cloud.google.com/http/host': hostname
		};

		var trace = tracing.startTrace();
		var xctc = req.get(XCTC_HEADER);
		if (xctc) {
			var match;
			match = xctc.match(TRACE_ID_REGEX);
			if (match && match.length === 2) trace.traceId = match[1];
			match = xctc.match(SPAN_ID_REGEX);
			if (match && match.length === 2) {
				var spanId = parseInt(match[1], 10);
				if (Number.isSafeInteger(spanId)) trace._spanIdInc = spanId;
			}
		}
		req.trace = trace.startSpan(url, labels);

		function endTrace() {
			// This is checked at the end so that consumers can still use the
			// req.trace property fully.
			if (!shouldTrace(req)) return;

			var statusCode = res.statusCode;
			if (statusCode !== undefined) statusCode = statusCode.toString();
			var contentLength = res.get('content-length');
			if (contentLength !== undefined) contentLength = contentLength.toString();

			var labels = {
				'trace.cloud.google.com/http/status_code': statusCode,
				'trace.cloud.google.com/http/response/size': contentLength
			};
			req.trace.end(labels);
			trace.end();
		}

		res.once('finish', endTrace);
		res.once('close', endTrace);

		next();
	};
};
